File: iteration2/src/course/FacultyTechnicalElectiveCourse.java
package iteration2.src.course;

import iteration2.src.human.Assistant;
import iteration2.src.human.Grade;
import iteration2.src.human.Lecturer;
import iteration2.src.human.Student;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class FacultyTechnicalElectiveCourse extends ElectiveCourse{

    public static final Map<Integer, Integer> numberOfCoursesTakeableBySemester = new HashMap<>(){{
        put(0,0);
        put(1,0);
        put(2,0);
        put(3,0);
        put(4,0);
        put(5,0);
        put(6,0);
        put(7,1);
    }};

    public FacultyTechnicalElectiveCourse(String code, String name, int credits, int theoreticalHours, int appliedHours,
                                          Grade firstYearToTake, Season firstSeasonToTake, List<Lecturer> lecturers, List<Assistant> assistants){
        super(code,name,credits,theoreticalHours,appliedHours,firstYearToTake,firstSeasonToTake,lecturers,assistants);
    }

    @Override
    public Boolean canStudentTakeCourse(Student student) {
        int semester = student.getStudentSemester();
        int noOfCoursesTillSemester = getTotalNumberOfCoursesUntilSemester(semester);

        return student.getTranscript().getNumberOfFTECoursesPassed() < noOfCoursesTillSemester
        && !student.didStudentPass(this);
    }

    public static int getTotalNumberOfCoursesUntilSemester(int semester){
        int noOfCoursesTillSemester = 0;

        for(int i = 0; i <= semester; i++){
            noOfCoursesTillSemester += numberOfCoursesTakeableBySemester.get(i).intValue();
        }

        return noOfCoursesTillSemester;
    }

    public int getCoursePriority(){
        return 1;
    }
}

File: iteration3/FacultyTechnicalElectiveCourse.py
class FacultyTechnicalElectiveCourse:

    number_of_courses_takeable_by_semester = {
        0:0,
        1:0,
        2:0,
        3:0,
        4:0,
        5:0,
        6:0,
        7:1
    }

    def __init__(self, code, name, credits, theoretical_hours, applied_hours, first_year_to_take, first_season_to_take, lecturers, assistants):
        super().__init__(code, name, credits, theoretical_hours, applied_hours, first_year_to_take, first_season_to_take, lecturers, assistants)

    def can_student_take_course(self, student):
        semester = student.get_student_semester()
        no_of_courses_till_semester = self.get_total_number_of_courses_until_semester(semester)

        return student.get_transcript().get_number_of_fte_courses_passed() < no_of_courses_till_semester and not student.did_student_pass(
            self)

    @staticmethod
    def get_total_number_of_courses_until_semester(semester):
        global number_of_courses_takeable_by_semester
        no_of_courses_till_semester = 0
        
        for i in range(semester + 1):
            no_of_courses_till_semester += number_of_courses_takeable_by_semester[i]

        return no_of_courses_till_semester

    def getCoursePriority(self):
        return 1